File: code/packages/typescript/vigenere-cipher/tests/vigenere-cipher.test.ts
import { describe, it, expect } from "vitest";
import {
  encrypt,
  decrypt,
  findKeyLength,
  findKey,
  breakCipher,
} from "../src/index.js";

// ---------------------------------------------------------------------------
// Long English text for cryptanalysis tests (~300+ chars of alpha content).
// This is required because IC analysis needs a statistically significant
// sample to reliably distinguish English from random text.
// ---------------------------------------------------------------------------
const LONG_ENGLISH_TEXT =
  "The quick brown fox jumps over the lazy dog and then runs around the " +
  "entire neighborhood looking for more adventures to embark upon while " +
  "the sun slowly sets behind the distant mountains casting long shadows " +
  "across the valley below where the river winds its way through ancient " +
  "forests filled with towering oak trees and singing birds that herald " +
  "the coming of spring with their melodious songs echoing through the " +
  "canopy above where squirrels chase each other from branch to branch " +
  "gathering acorns and other nuts for the long winter months ahead when " +
  "the ground will be covered in a thick blanket of pristine white snow " +
  "and the children will build snowmen and throw snowballs at each other " +
  "laughing and playing until their parents call them inside for dinner " +
  "where warm soup and fresh bread await them on the old wooden table";

// ---------------------------------------------------------------------------
// Encrypt tests
// ---------------------------------------------------------------------------
describe("encrypt", () => {
  it("encrypts ATTACKATDAWN with key LEMON", () => {
    expect(encrypt("ATTACKATDAWN", "LEMON")).toBe("LXFOPVEFRNHR");
  });

  it("preserves case and punctuation", () => {
    expect(encrypt("Hello, World!", "key")).toBe("Rijvs, Uyvjn!");
  });

  it("handles all-lowercase text", () => {
    expect(encrypt("attackatdawn", "lemon")).toBe("lxfopvefrnhr");
  });

  it("handles mixed case key", () => {
    // Key is case-insensitive: "LeMoN" should work the same as "LEMON"
    expect(encrypt("ATTACKATDAWN", "LeMoN")).toBe("LXFOPVEFRNHR");
  });

  it("handles single-char key", () => {
    // Key "B" means shift of 1 for every letter (same as Caesar +1)
    expect(encrypt("ABC", "B")).toBe("BCD");
  });

  it("handles key longer than alpha content", () => {
    // A shifted by L(11) = L, B shifted by O(14) = P
    expect(encrypt("AB", "LONGERKEY")).toBe("LP");
  });

  it("skips non-alpha for key advancement", () => {
    // "A T" -- space does not advance key
    // A shifted by L(11) = L, space passes through, T shifted by E(4) = X
    expect(encrypt("A T", "LE")).toBe("L X");
  });

  it("handles digits and special chars unchanged", () => {
    expect(encrypt("Hello 123!", "key")).toBe("Rijvs 123!");
  });

  it("handles empty plaintext", () => {
    expect(encrypt("", "key")).toBe("");
  });

  it("throws on empty key", () => {
    expect(() => encrypt("hello", "")).toThrow("Key must not be empty");
  });

  it("throws on non-alpha key", () => {
    expect(() => encrypt("hello", "key1")).toThrow(
      "Key must contain only alphabetic characters",
    );
    expect(() => encrypt("hello", "ke y")).toThrow(
      "Key must contain only alphabetic characters",
    );
  });
});

// ---------------------------------------------------------------------------
// Decrypt tests
// ---------------------------------------------------------------------------
describe("decrypt", () => {
  it("decrypts LXFOPVEFRNHR with key LEMON", () => {
    expect(decrypt("LXFOPVEFRNHR", "LEMON")).toBe("ATTACKATDAWN");
  });

  it("preserves case and punctuation", () => {
    expect(decrypt("Rijvs, Uyvjn!", "key")).toBe("Hello, World!");
  });

  it("handles all-lowercase", () => {
    expect(decrypt("lxfopvefrnhr", "lemon")).toBe("attackatdawn");
  });

  it("handles empty ciphertext", () => {
    expect(decrypt("", "key")).toBe("");
  });

  it("throws on empty key", () => {
    expect(() => decrypt("hello", "")).toThrow("Key must not be empty");
  });

  it("throws on non-alpha key", () => {
    expect(() => decrypt("hello", "123")).toThrow(
      "Key must contain only alphabetic characters",
    );
  });
});

// ---------------------------------------------------------------------------
// Round-trip tests
// ---------------------------------------------------------------------------
describe("round-trip", () => {
  const cases = [
    { text: "ATTACKATDAWN", key: "LEMON" },
    { text: "Hello, World!", key: "key" },
    { text: "The quick brown fox!", key: "SECRET" },
    { text: "abc def ghi", key: "xyz" },
    { text: "MiXeD CaSe 123", key: "AbCdE" },
    { text: "a", key: "z" },
    { text: "ZZZZZZ", key: "A" },
  ];

  for (const { text, key } of cases) {
    it(`decrypt(encrypt("${text}", "${key}"), "${key}") == original`, () => {
      expect(decrypt(encrypt(text, key), key)).toBe(text);
    });
  }
});

// ---------------------------------------------------------------------------
// Cryptanalysis: findKeyLength
// ---------------------------------------------------------------------------
describe("findKeyLength", () => {
  it("finds key length for text encrypted with a 5-letter key", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "LEMON");
    const length = findKeyLength(ct);
    expect(length).toBe(5);
  });

  it("finds key length for text encrypted with a 6-letter key", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "SECRET");
    const length = findKeyLength(ct);
    expect(length).toBe(6);
  });

  it("finds key length for text encrypted with a 3-letter key", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "KEY");
    const length = findKeyLength(ct);
    expect(length).toBe(3);
  });

  it("respects maxLength parameter", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "LEMON");
    // With maxLength=3, cannot find length 5 -- returns best of 2,3
    const length = findKeyLength(ct, 3);
    expect(length).toBeGreaterThanOrEqual(1);
    expect(length).toBeLessThanOrEqual(3);
  });

  it("returns 1 for very short text", () => {
    expect(findKeyLength("A")).toBe(1);
  });
});

// ---------------------------------------------------------------------------
// Cryptanalysis: findKey
// ---------------------------------------------------------------------------
describe("findKey", () => {
  it("recovers LEMON from ciphertext with known key length 5", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "LEMON");
    const key = findKey(ct, 5);
    expect(key).toBe("LEMON");
  });

  it("recovers SECRET from ciphertext with known key length 6", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "SECRET");
    const key = findKey(ct, 6);
    expect(key).toBe("SECRET");
  });

  it("recovers KEY from ciphertext with known key length 3", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "KEY");
    const key = findKey(ct, 3);
    expect(key).toBe("KEY");
  });
});

// ---------------------------------------------------------------------------
// Cryptanalysis: breakCipher
// ---------------------------------------------------------------------------
describe("breakCipher", () => {
  it("automatically breaks a Vigenere cipher with key LEMON", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "LEMON");
    const result = breakCipher(ct);
    expect(result.key).toBe("LEMON");
    expect(result.plaintext).toBe(LONG_ENGLISH_TEXT);
  });

  it("automatically breaks a Vigenere cipher with key SECRET", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "SECRET");
    const result = breakCipher(ct);
    expect(result.key).toBe("SECRET");
    expect(result.plaintext).toBe(LONG_ENGLISH_TEXT);
  });

  it("recovered plaintext matches original", () => {
    const ct = encrypt(LONG_ENGLISH_TEXT, "CIPHER");
    const result = breakCipher(ct);
    // Even if key recovery isn't perfect, decrypting with the recovered
    // key should be internally consistent
    expect(decrypt(encrypt(LONG_ENGLISH_TEXT, result.key), result.key)).toBe(
      LONG_ENGLISH_TEXT,
    );
  });
});

File: code/packages/cpp/vigenere-cipher/tests/vigenere_cipher_test.cpp
// Tests for the C++ vigenere-cipher, using the iso_test.h harness. The cipher
// vectors and cryptanalysis cases are taken from the Rust crate's own tests.
#include "iso_test.h"

#include <string>

#include "vigenere_cipher.hpp"

namespace vig = ca::vigenere;

// The crate's long English sample for cryptanalysis.
static const std::string LONG_ENGLISH_TEXT =
    "The quick brown fox jumps over the lazy dog and then runs around the "
    "entire neighborhood looking for more adventures to embark upon while "
    "the sun slowly sets behind the distant mountains casting long shadows "
    "across the valley below where the river winds its way through ancient "
    "forests filled with towering oak trees and singing birds that herald "
    "the coming of spring with their melodious songs echoing through the "
    "canopy above where squirrels chase each other from branch to branch "
    "gathering acorns and other nuts for the long winter months ahead when "
    "the ground will be covered in a thick blanket of pristine white snow "
    "and the children will build snowmen and throw snowballs at each other "
    "laughing and playing until their parents call them inside for dinner "
    "where warm soup and fresh bread await them on the old wooden table";

static void check_encrypt(const std::string& pt, const std::string& key,
                          const std::string& expected) {
    auto got = vig::encrypt(pt, key);
    ISO_CHECK(got.has_value());
    if (got) {
        ISO_CHECK(*got == expected);
    }
}

static void check_decrypt(const std::string& ct, const std::string& key,
                          const std::string& expected) {
    auto got = vig::decrypt(ct, key);
    ISO_CHECK(got.has_value());
    if (got) {
        ISO_CHECK(*got == expected);
    }
}

int main() {
    // Encrypt vectors (from cipher.rs tests).
    check_encrypt("ATTACKATDAWN", "LEMON", "LXFOPVEFRNHR");
    check_encrypt("Hello, World!", "key", "Rijvs, Uyvjn!");
    check_encrypt("attackatdawn", "lemon", "lxfopvefrnhr");
    check_encrypt("ATTACKATDAWN", "LeMoN", "LXFOPVEFRNHR");
    check_encrypt("ABC", "B", "BCD");
    check_encrypt("A T", "LE", "L X");
    check_encrypt("Hello 123!", "key", "Rijvs 123!");
    check_encrypt("", "key", "");

    // Decrypt vectors.
    check_decrypt("LXFOPVEFRNHR", "LEMON", "ATTACKATDAWN");
    check_decrypt("Rijvs, Uyvjn!", "key", "Hello, World!");
    check_decrypt("lxfopvefrnhr", "lemon", "attackatdawn");
    check_decrypt("", "key", "");

    // Invalid keys are rejected (nullopt).
    ISO_CHECK(!vig::key_valid(""));
    ISO_CHECK(!vig::key_valid("key1"));
    ISO_CHECK(!vig::key_valid("ke y"));
    ISO_CHECK(vig::key_valid("LEMON"));
    ISO_CHECK(!vig::encrypt("hello", "").has_value());
    ISO_CHECK(!vig::encrypt("hello", "key1").has_value());
    ISO_CHECK(!vig::decrypt("hello", "123").has_value());

    // Round trips.
    {
        const char* texts[] = {"ATTACKATDAWN", "Hello, World!",
                               "The quick brown fox!", "MiXeD CaSe 123",
                               "ZZZZZZ"};
        const char* keys[] = {"LEMON", "key", "SECRET", "AbCdE", "A"};
        for (int i = 0; i < 5; ++i) {
            auto ct = vig::encrypt(texts[i], keys[i]);
            ISO_CHECK(ct.has_value());
            auto pt = vig::decrypt(*ct, keys[i]);
            ISO_CHECK(pt.has_value());
            ISO_CHECK(*pt == std::string(texts[i]));
        }
    }

    // Cryptanalysis: key-length detection.
    {
        auto ct5 = vig::encrypt(LONG_ENGLISH_TEXT, "LEMON");
        auto ct6 = vig::encrypt(LONG_ENGLISH_TEXT, "SECRET");
        auto ct3 = vig::encrypt(LONG_ENGLISH_TEXT, "KEY");
        ISO_CHECK_EQ_UINT(vig::find_key_length(*ct5, 20), 5u);
        ISO_CHECK_EQ_UINT(vig::find_key_length(*ct6, 20), 6u);
        ISO_CHECK_EQ_UINT(vig::find_key_length(*ct3, 20), 3u);
        ISO_CHECK_EQ_UINT(vig::find_key_length("A", 20), 1u);
    }

    // Cryptanalysis: key recovery.
    {
        auto ct5 = vig::encrypt(LONG_ENGLISH_TEXT, "LEMON");
        auto ct6 = vig::encrypt(LONG_ENGLISH_TEXT, "SECRET");
        auto ct3 = vig::encrypt(LONG_ENGLISH_TEXT, "KEY");
        ISO_CHECK(vig::find_key(*ct5, 5) == "LEMON");
        ISO_CHECK(vig::find_key(*ct6, 6) == "SECRET");
        ISO_CHECK(vig::find_key(*ct3, 3) == "KEY");
    }

    // Full automatic break.
    {
        auto ct = vig::encrypt(LONG_ENGLISH_TEXT, "LEMON");
        vig::BreakResult result = vig::break_cipher(*ct);
        ISO_CHECK(result.key == "LEMON");
        ISO_CHECK(result.plaintext == LONG_ENGLISH_TEXT);
    }

    // Break is at least self-consistent for the key it recovers.
    {
        auto ct = vig::encrypt(LONG_ENGLISH_TEXT, "CIPHER");
        vig::BreakResult result = vig::break_cipher(*ct);
        auto re = vig::encrypt(LONG_ENGLISH_TEXT, result.key);
        auto rt = vig::decrypt(*re, result.key);
        ISO_CHECK(*rt == LONG_ENGLISH_TEXT);
    }

    return ISO_TEST_RESULT();
}
